packages/diagram-ui/src/features/toolbox/toolbox.ts

```diff
@@ -0,0 +1,548 @@
+    }
+
+    /**
+     * Generates the search input element.
+     *
+     * @returns The search input element
+     */
+    private generateSearchInput() {
+        return h("input", {
+            props: {
+                placeholder: "Search"
+            },
+            hook: {
+                insert: (vnode) => {
+                    const element = vnode.elm as HTMLInputElement;
+                    element.focus();
+                }
+            },
+            on: {
+                input: (event) => {
+                    this.searchString = (event.target as HTMLInputElement).value;
+                    this.update();
+                },
+                keydown: (event) => {
+                    if (event.key === "Escape") {
+                        this.searchString = undefined;
+                        this.update();
+                    }
+                }
+            }
+        });
+    }
+
+    /**
+     * Toggles the toolbox.
+     */
+    private toggleToolbox(): void {
+        this.isClosed = !this.isClosed;
+        this.update();
+        this.configManager.updateConfig({ toolboxDisabled: this.isClosed });
+    }
+
+    /**
+     * Generates a codicon button.
+     *
+     * @param icon the name of the codicon
+     * @param title the title of the button
+     * @param action the action to execute when the button is clicked
+     * @param active if true, the active class is added
+     * @returns the codicon button
+     */
+    private generateCodiconButton(icon: string, title: string, action: () => void, active: boolean = false): VNode {
+        return h(
+            "button.codicon-button",
+            {
+                on: {
+                    click: action
+                },
+                class: {
+                    active
+                }
+            },
+            [this.generateCodicon(icon, title)]
+        );
+    }
+
+    /**
+     * Generates a codicon.
+     *
+     * @param icon the name of the codicon
+     * @param title the title of the codicon
+     * @returns the codicon
+     */
+    private generateCodicon(icon: string, title: string): VNode {
+        return h("i.codicon", {
+            attrs: {
+                title
+            },
+            class: {
+                [`codicon-${icon}`]: true
+            }
+        });
+    }
+
+    /**
+     * Generates the UI for the toolbox items.
+     *
+     * @param root The root element providing the toolbox edits
+     * @returns The UI for the toolbox items
+     */
+    private generateToolboxItems(root: Root): VNode[] | VNode {
+        if (this.searchString != undefined && this.searchString.length > 0) {
+            return this.generateFilteredToolboxItems(this.searchString);
+        } else {
+            return this.generateGroupedToolboxItems(root);
+        }
+    }
+
+    /**
+     * Generates the UI for the toolbox items grouped by group.
+     * Used if no search is active.
```

Comment:
Wait, we don't show the groups while searching anymore?

Reply:
The order is determined by the ranking from the full text search index. Thus, entries are no longer grouped by their group. Do you have any good UI ideas for displaying the group?

Reply:
My idea for displaying the UI would have been to simply hide any entry within any group, if the search text does not match it.
And if there are no entries left to show for a group, I would have hidden the group.

Reply:
the problem with this is, we use a full text index with fuzzy search
the quality of the match can vary highly between the highest and the lowest match
for that reason, you usually sort in that order, and not just filter the results

Reply:
the mean is not stable against outliers
imagine if you have a perfect match in one group
however also some rather bad matches
and then you have several other groups where you have ok matches,
then the best match would be displayed rather low, because the bad matches in that group drag it down

Reply:
I'm also fine with using an adjusted median: Use the element in the middle, or the one at `middle + 1` when there are even options per group.
But more importantly: What are your thoughts on the suggested process in the first place?